src/core_plugins/tag_cloud/public/tag_cloud.js

```diff
@@ -0,0 +1,250 @@
+  }
+
+
+  setSize(newSize) {
+    if (newSize[0] === 0 || newSize[1] === 0) {
+      return;
+    }
+    if (newSize[0] === this._size[0] && newSize[1] === this._size[1]) {
+      return;
+    }
+
+    this._size = newSize;
+
+
+    if (this._complete && this._cloudWidth < this._size[0] && this._cloudHeight < this._size[1]) {
+      this._updateContainerSize();
+      return;
+    }
+
+    this._washWords();
+    this._invalidate();
+
+  }
+
+  setData(data) {
+    this._words = data.map(toWordTag);
+    this._makeTextSizeMapper();
+    this._invalidate();
+  }
+
+
+  destroy() {
+    cancelAnimationFrame(this._domManipulationFrame);
+    clearTimeout(this._timeoutHandle);
+    this._element.innerHTML = '';
+  }
+
+
+  _updateContainerSize() {
+    this._d3SvgContainer.attr('width', this._size[0]);
+    this._d3SvgContainer.attr('height', this._size[1]);
+    this._svgGroup.attr('width', this._size[0]);
+    this._svgGroup.attr('height', this._size[1]);
+    this._svgGroup.attr('transform', 'translate(' + this._size[0] / 2 + ',' + this._size[1] / 2 + ')');
+  }
+
+  _washWords() {
+    if (!this._words) {
+      return;
+    }
+
+    //the tagCloudLayoutGenerator clobbers the word-object with metadata about positioning.
+    //This can causes corrupt states in the layout-generator
+    //where words get collapsed to the same location and do not reposition correctly.
+    //=> we recreate an empty word object without the metadata
+    this._words = this._words.map(toWordTag);
+    this._makeTextSizeMapper();
+  }
+
+  _onLayoutEnd(resolve, reject, wordsWithLayout) {
+    this._domManipulationFrame = null;
+    const svgTextNodes = this._svgGroup.selectAll('text');
+    const stage = svgTextNodes.data(wordsWithLayout, getText);
+
+    const enterSelection = stage.enter();
+    const enteringTags = enterSelection.append('text');
+    enteringTags.style('font-size', getSizeInPixels);
+    enteringTags.style('font-style', this._fontStyle);
+    enteringTags.style('font-weight', () => this._fontWeight);
+    enteringTags.style('font-family', () => this._fontFamily);
+    enteringTags.style('fill', getFill);
+    enteringTags.attr('text-anchor', () => 'middle');
+    enteringTags.attr('transform', positionWord);
+    enteringTags.text(getText);
+
+    const movingTags = stage.transition();
+    movingTags.duration(600);
+    movingTags.style('font-size', getSizeInPixels);
+    movingTags.attr('transform', positionWord);
+    movingTags.style('fill-opacity', 1);
+
+    const exitingTags = stage.exit();
+    const exitTransition = exitingTags.transition();
+    exitTransition.duration(200);
+    exitingTags.style('fill-opacity', 1e-6);
+    exitingTags.attr('font-size', 1);
+    exitingTags.remove();
+
+    let exits = 0;
+    let moves = 0;
+    const resolveWhenDone = () => {
+      if (exits === 0 && moves === 0) {
+        resolve(true);
+        //fire rendered event here..
+      }
+    };
+    exitTransition.each(_ => exits += 1);
```

Comment:
I'm not absolutely sure but can this just be:

```javascript
exits += exitTransition.nodes().length;
```

Or at the very least we can remove the `_` argument and use an increment operator:

```javascript
exitTransition.each(() => exits++);
```